File: Python3/_206.py
# 206: Reverse linked list
# Track 3 pointers: back; lead: back.next before swap;
# nextLead: record lead.next before swap. Make lead
# point to back, shift back and lead up for next iter.

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if not head or not head.next:
            return head

        back = head
        lead = back.next

        while lead: # If lead, guarantees back is non-None
            # Record before rewiring would overwrite this data
            nextLead = lead.next

            lead.next = back
            back, lead = lead, nextLead # lead takes role of next back

        head.next = None

        # "while lead" exited on back that is tail of original
        # list (ie next == None). Put it as head of reversed list
        return back 

File: Cpp/_206.cpp
/* 206: Reverse linked list
 * Track 3 pointers: back: nearest to head; lead: back.next 
 * before swap; nextLead: lead.next before swap. Rewire 
 * lead.next = back; shift back, lead up for next iter */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;

        ListNode* back = head;
        ListNode* lead = back->next;

        // If lead, guarantees back is non-None
        while (lead) {
            // Record before rewiring would overwrite this data
            ListNode* nextBack = lead->next;

            lead->next = back;

            back = lead; // lead takes role of next back
            lead = nextBack; 
        }

        head->next = nullptr;

        // "while lead" exited on back that is tail of original
        // list (ie next == null). Put it as head of reversed list
        return back;
    }
};